scripts/prune-environment.py

```diff
@@ -19,24 +21,50 @@
 # Only pin explicit dependencies
 
 
-def iterate_deps(xs, ys, accumulator):
-    for x in xs:
-        for y in ys:
-            if x.split("=")[0] == y.split("=")[0]:
-                accumulator.append(x)
+def iterate_deps(
+    pinned_deps: Iterable[str], unpinned_deps: Iterable[str], accumulator: List[str]
+) -> None:
+    for pinned_dep in pinned_deps:
+        for unpinned_dep in unpinned_deps:
+            for sep in ["=", "<", ">"]:
+                if pinned_dep.split("=")[0] == unpinned_dep.split(sep)[0]:
+                    accumulator.append(pinned_dep)
+                    break  # break out of the sep loop to avoid duplicate entries
+
 
+deps_not_pip: List[str] = []
+deps_pip: List[str] = []
 
-deps_not_pip = []
-deps_pip = []
+env_unpinned_contains_pip = "pip" in env_unpinned["dependencies"][-1]
+env_unpinned_dev_contains_pip = "pip" in env_unpinned_dev["dependencies"][-1]
 iterate_deps(
     env_pinned["dependencies"][:-1],
-    env_unpinned["dependencies"][:-1] + env_unpinned_dev["dependencies"][:-1],
+    (
+        env_unpinned["dependencies"][:-1]
+        if env_unpinned_contains_pip
+        else env_unpinned["dependencies"]
+    )
+    + (
+        env_unpinned_dev["dependencies"][:-1]
+        if env_unpinned_dev_contains_pip
+        else env_unpinned_dev["dependencies"]
+    ),
     deps_not_pip,
 )
+
+# Checking if there are any pip dependencies
+try:
+    deps_pip_unpinned = env_unpinned["dependencies"][-1]["pip"]
+except:
+    deps_pip_unpinned = []
+try:
+    deps_pip_unpinned_dev = env_unpinned_dev["dependencies"][-1]["pip"]
+except:
+    deps_pip_unpinned_dev = []
```

Comment:
Just a suggestion, maybe we could reuse `env_unpinned_contains_pip` and `env_unpinned_dev_contains_pip` here instead of checking again in a different way